swupd/rename.go

```diff
@@ -0,0 +1,185 @@
+		}
+	}
+}
+
+// linkrenamepair links two files together
+func linkrenamepair(a, b *File) {
+	a.DeltaPeer = b
+	b.DeltaPeer = a
+	a.Rename = true
+	b.Rename = true
+}
+
+// trimrenamed returns an slice which has had files with DeltaPeers purged
+func trimrenamed(a []*File) []*File {
+	r := make([]*File, 0, len(a))
+	for _, f := range a {
+		// Do not worry about ghosted files here, assume that they will
+		// exist on the target system, and so can be used as the source for
+		// renames. The classic ghosted files are the OS kernels and modules,
+		// and it is helpful to be able to ship just the differences for them.
+		// Worst case is that the files do not exist on the target system, in which
+		// case the swupd-client will fall back to doing a full download.
+		if f.DeltaPeer == nil {
+			r = append(r, f)
+		}
+	}
+	return r
+}
+
+// trimsmall returns an slice which has had files with small files removed
+func trimsmall(a []*File, minsize int64) []*File {
+	r := make([]*File, 0, len(a))
+	for _, f := range a {
+		if f.Info == nil {
+			panic("Missing f.Info for " + f.Name)
+		}
+		if f.Info.Size() > minsize {
+			r = append(r, f)
+		}
+	}
+	return r
+}
+
+// pairednames holds a *File and another "name", which is used for various
+// alterations of the filename for matching purposes. In particular removing
+// digits so a new release of a shared library is matched to previous ones
+type pairednames struct {
+	f *File
+	n string // Filename with digits removed
+}
+
+// stripdigits is used with strings.Map to remove digits from filename
+func stripdigits(r rune) rune {
+	if r >= '0' && r <= '9' {
+		return -1 // strings.Map removes negative values
+	}
+	return r
+}
+
+// makePairedNames returns `a` sorted by name, disregarding digits.
+// The secondary sort key is the original name. This is mainly intended for
+// cases like with python we have the same filename under both python3.6 and python2.7
+// it would look weird to rename the old python2.7/file1 to python3.6/file2 and at
+// the same time rename the old python3.6/file1 to python2.7/file2.
+func makePairedNames(list []*File) []pairednames {
+	pairs := make([]pairednames, len(list))
+	for i, f := range list {
+		pairs[i].f = f
+		pairs[i].n = strings.Map(stripdigits, f.Name)
+	}
+	sort.Slice(pairs, func(a, b int) bool {
+		if pairs[a].n == pairs[b].n { // Same stripped name, sort on original name
+			// mainly for python2.x vs 3.x
+			return pairs[a].f.Name < pairs[b].f.Name
+		}
+		return pairs[a].n < pairs[b].n
+	})
+	return pairs
+}
+
+// fixupstatfields adds the missing stat fields
+// construct the path to the old chroot by Joining the c.imageBase
+// file.Previous, bundle name, and file.Name fields
+func fixupstatfields(needed []*File, m *Manifest, c *config) error {
+	var bundleChroot string
+	for i := range needed {
+		if needed[i].Info != nil {
+			continue
+		}
+		if bundleChroot == "" {
+			bundleChroot = filepath.Join(c.imageBase, fmt.Sprint(m.Header.Previous), m.Name)
```

Comment:
The panic happens here because the bundle-chroot does not exist anymore when this happens, and only "full" is available. This should always use "full" instead of m.Name since not everyone will build with --keep-chroots flag on, so there won't be a dir to stat() anymore. :raised_hands:

Reply:
This appears to be resolved. No more panics for me.

Reply:
Just did what @tmarcu told me to do!

Just pushing a rebase to fix the delta.go conflict